lib/workflow/stores/mongo.js

```diff
@@ -0,0 +1,563 @@
+            graphId: graphId,
+            state: Constants.TaskStates.Pending,
+            dependencies: taskDependencyItem.dependencies,
+            terminalOnStates: taskDependencyItem.terminalOnStates
+        };
+
+        return waterline.taskdependencies.create(obj);
+    };
+
+    exports.getTaskById = function(data) {
+        assert.object(data, 'data');
+        assert.uuid(data.graphId, 'data.graphId');
+        assert.uuid(data.taskId, 'data.taskId');
+
+        var query = {
+            instanceId: data.graphId
+        };
+        var options = {
+            fields: {
+                _id: 0,
+                instanceId: 1,
+                context: 1,
+                tasks: {}
+            }
+        };
+        options.fields.tasks[data.taskId] = 1;
+
+        return waterline.graphobjects.findOne(query, options)
+        .then(function(graph) {
+            return {
+                graphId: graph.instanceId,
+                context: graph.context,
+                task: graph.tasks[data.taskId]
+            };
+        });
+    };
+
+    exports.getServiceGraphs = function() {
+        var query = {
+            serviceGraph: true
+        };
+
+        return waterline.graphobjects.find(query)
+        .then(function(graphs) {
+            return _.map(graphs, function(graph) {
+                return graph.toJSON();
+            });
+        });
+    };
+
+    exports.heartbeatTasksForRunner = function(leaseId) {
+        assert.uuid(leaseId, 'leaseId');
+
+        var query = {
+            taskRunnerLease: leaseId,
+            reachable: true,
+            state: Constants.TaskStates.Pending
+        };
+        var update = {
+            $set: {
+                taskRunnerLease: leaseId,
+                taskRunnerHeartbeat: new Date()
+            }
+        };
+        var options = {
+            multi: true
+        };
+
+        return waterline.taskdependencies.updateMongo(query, update, options);
+    };
+
+    exports.getOwnTasks = function(leaseId) {
+        assert.uuid(leaseId, 'leaseId');
+
+        var query = {
+            where: {
+                taskRunnerLease: leaseId,
+                reachable: true,
+                state: Constants.TaskStates.Pending
+            }
+        };
+
+        return waterline.taskdependencies.find(query);
+    };
+
+    exports.findActiveGraphForTarget = function(target) {
+        if (!target) {
+            return Promise.resolve(null);
+        }
+        assert.string(target, 'target');
+
+        var query = {
+            "context.target": target,
+            _status: Constants.TaskStates.Pending
+        };
+
+        return waterline.graphobjects.findOneMongo(query);
+    };
+
+    exports.findActiveGraphs = function(domain) {
+        assert.string(domain, 'domain');
+
+        var query = {
+            domain: domain,
+            _status: Constants.TaskStates.Pending
+        };
+
+        return waterline.graphobjects.find(query);
```

Comment:
There are some conveniences to the response objects that come out of waterline too, so we like to use them where we can.